airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_with_global.py

```diff
@@ -0,0 +1,179 @@
+
+
+class PerPartitionWithGlobalCursor(DeclarativeCursor):
+    """
+    Manages state for streams with multiple partitions, with an optional fallback to a global cursor when specific conditions are met.
+
+    This cursor handles partitioned streams by maintaining individual state per partition using `PerPartitionCursor`. If the number of partitions exceeds a defined limit, it switches to a global cursor (`GlobalSubstreamCursor`) to manage state more efficiently.
+
+    **Overview**
+
+    - **Partition-Based State**: Initially manages state per partition to ensure accurate processing of each partition's data.
+    - **Global Fallback**: Switches to a global cursor when the partition limit is exceeded to handle state management more effectively.
+
+    **Switching Logic**
+
+    - Monitors the number of partitions.
+    - If `PerPartitionCursor.limit_reached()` returns `True`, sets `_use_global_cursor` to `True`, activating the global cursor.
+
+    **Active Cursor Selection**
+
+    - Uses the `_get_active_cursor()` helper method to select the active cursor based on the `_use_global_cursor` flag.
+    - This simplifies the logic and ensures consistent cursor usage across methods.
+
+    **State Structure Example**
+
+    ```json
+    {
+        "states": [
+            {
+                "partition": {"partition_key": "partition_1"},
+                "cursor": {"cursor_field": "2021-01-15"}
+            },
+            {
+                "partition": {"partition_key": "partition_2"},
+                "cursor": {"cursor_field": "2021-02-14"}
+            }
+        ],
+        "state": {
+            "cursor_field": "2021-02-15"
+        },
+        "use_global_cursor": false
+    }
+    ```
+
+    In this example, the cursor is using partition-based state management (`"use_global_cursor": false`), maintaining separate cursor states for each partition.
+
+    **Usage Scenario**
+
+    Suitable for streams where the number of partitions may vary significantly, requiring dynamic switching between per-partition and global state management to ensure data consistency and efficient synchronization.
+    """
+
+    def __init__(self, cursor_factory: CursorFactory, partition_router: PartitionRouter, stream_cursor: DatetimeBasedCursor):
+        self._partition_router = partition_router
+        self._per_partition_cursor = PerPartitionCursor(cursor_factory, partition_router)
+        self._global_cursor = GlobalSubstreamCursor(stream_cursor, partition_router)
+        self._use_global_cursor = False
+        self._current_partition: Optional[Mapping[str, Any]] = None
+        self._last_slice: bool = False
+
+    def _get_active_cursor(self) -> Union[PerPartitionCursor, GlobalSubstreamCursor]:
+        return self._global_cursor if self._use_global_cursor else self._per_partition_cursor
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        self._global_cursor.start_slices_generation()
```

Comment:
I had a thought during the night: Do we expect to have `GlobalSubstreamCursor` without `PerPartitionCursor` for a connector? i.e. that a connector would just do `GlobalSubstreamCursor`? 

When I read this code yesterday, I feared that we need to expose a lot of internal logic of global substream cursors. What makes me think that? 
* We need to make some things public like `start_slices_generation`, `generation_slices_from_partition` and `register_slice`
* The loop here is a bit similar to `GlobalSubstreamCursor.stream_slices`
It feels like all of this could be hidden if everything was part of one class. Would it make sense to have `GlobalSubstreamCursor` have the `PerPartitionCursor`. If we eventually want just a `GlobalSubstreamCursor` with `PerPartitionCursor`, we could always make it nullable or something.

Reply:
Yes, the `GlobalSubstreamCursor` can be used independently by enabling the `global_substream_cursor` option.

As for combining the classes, I don't think it's the best approach since they handle cursor state differently. Merging them would introduce a lot of conditional logic (e.g., for closing slices), making the code harder to maintain and reducing the clarity of using separate classes.

Reply:
Ok! I'm fine with that. Anyway, we will probably move this cursor to support concurrency next cycle which we can re-evaluate at that point how maintainable this is